app/gui2/src/components/GraphEditor/widgets/WidgetFileBrowser.vue

```diff
@@ -0,0 +1,126 @@
+<script setup lang="ts">
+import NodeWidget from '@/components/GraphEditor/NodeWidget.vue'
+import { asNot } from '@/util/data/types.ts'
+import { provideCustomDropdownItems } from '@/providers/customDropdownItems'
+import { defineWidget, Score, WidgetInput, widgetProps } from '@/providers/widgetRegistry'
+import type { TokenId } from '@/util/ast/abstract.ts'
+import { ArgumentInfoKey } from '@/util/callTree'
+import { computed, ref } from 'vue'
+import { Ast } from '@/util/ast'
+import { useGraphStore } from '@/stores/graph'
+import type { RequiredImport } from '@/stores/graph/imports'
+
+const props = defineProps(widgetProps(widgetDefinition))
+const graph = useGraphStore()
+
+const insertAsFileConstructor = computed(() => {
+  const reprType = props.input[ArgumentInfoKey]?.info?.reprType
+  if (reprType == null) return false
+  const textTypePos = reprType.indexOf(TEXT_TYPE)
+  const fileTypePos = reprType.indexOf(FILE_TYPE)
+  // If both types are present, check if TEXT_TYPE comes after FILE_TYPE.
+  if (fileTypePos !== -1 && (textTypePos === -1 || textTypePos > fileTypePos)) {
+    return true
+  } else {
+    return false
+  }
+})
+const strictlyFile = computed(() => props.input.dynamicConfig?.kind === 'File_Browse')
+const strictlyDirectory = computed(() => props.input.dynamicConfig?.kind === 'Folder_Browse')
+const label = computed(() => strictlyDirectory.value ? 'Choose directory…' : 'Choose file…')
+
+const FILE_CONSTRUCTOR = FILE_TYPE + '.new'
+const FILE_SHORT_CONSTRUCTOR = 'File.new'
+
+function makeValue(edit: Ast.MutableModule, useFileConstructor: boolean, path: string): Ast.Owned {
+  if (useFileConstructor) {
+    const arg = Ast.TextLiteral.new(path, edit)
+    const requiredImport = {
+      kind: 'Unqualified',
+      from: FILE_MODULE,
+      import: 'File'
+    } as RequiredImport
+    const conflicts = graph.addMissingImports(edit, [requiredImport])
+    const constructor = conflicts != null ? FILE_CONSTRUCTOR : FILE_SHORT_CONSTRUCTOR
+    return Ast.parse(`${constructor} ${arg.code()}`, edit)
```

Comment:
It is, though using it here complicates things without real benefit. We’re carefully constructing the string in such a way that the contents is very predictable: it would always be a qualified name + already escaped string literal. Using `Ast.App.new` would require constructing the qualified name separately, and again from string (what it would be? Property access chain? I don’t see any API available for that).

Reply:
The benefit is not needing to do anything "carefully". I'm sure it's correct now, but later we can make changes more easily if the types guarantee correctness than if we have to be careful about escaping and precedence in the concatenated string.

There's no helper yet combining `qnSegments` with `PropertyAccess.Sequence`--that would be a useful one-liner if you want to write one. Otherwise, if type-safe code construction would add too much time to this PR, please at least use `Ast.App.tryParse`. `Ast.parse` is only for handling arbitrary Enso code without any expectation about the content.

Reply:
Sure, fixed.